### PR TITLE
Use a centralised logger interface

### PR DESCRIPTION
**Pull request by SimonRichardson**
This standardizes all of the logging in juju to one and only one interface. This will ensure that when we implement status history we can guarantee that the logging is correctly nested and can pass labels through.

This is a lot of work to push this all through. This PR is all about swapping out the direct calls to loggo and using the core logger package Logger interface. The backing for the Logger interface is still loggo. There should be not change to logging in juju, this refactoring provides numerous advantages:

 1. Standardize logging, there is only one way to do it.
 2. As `loggo.Logger` was self-referential (type refers to itself: Child as an example) it was becoming difficult to use when each package had its own logger interface.
 3. Testing becomes simpler, by using CheckLog, we can just write to the testing context. No loggo is involved. 
 4. Removal of over ~1.7k~ ~2.1k~ 5k lines of code, which reduces complexity.

The disadvantage is that once you pull on a thread, you have to change it all.

### Future work

 1. Context will be added to all logging methods in the future. Giving the option of linking traces to logs and vica-versa. 
 2. Each model logger will have the model UUID associated with the model logger, removing the logsink in the modelworker. The writing of model logs will then be written to the right logs based on the record, thus simplifying the logging.
 3. Status history can then be written to a log, and it will be written to the correct model log based on 2.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd test --debug
$ juju debug-log -m controller
$ juju add-model default
$ juju deploy ubuntu
$ juju debug-log
```

## Links

**Jira card:** JUJU-5956

### CODE REVIEW

File: core/logger/interface.go [hunk 1]
```diff
@@ -58,6 +152,59 @@ type Logger interface {
...
+
+	// AddWriter adds a writer to the list to be called for each logging call.
+	// The name cannot be empty, and the writer cannot be nil. If an existing
+	// writer exists with the specified name, an error is returned.
+	//
+	// Note: we're relying on loggo.Writer here, until we do model level logging.
+	// Deprecated: This will be removed in the future and is only here whilst
+	// we cut things across.
+	AddWriter(name string, writer loggo.Writer) error
 }
```
Comment (SimonRichardson) [inline core/logger/interface.go:208]: This

File: internal/logger/loggo.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
+// Copyright 2024 Canonical Ltd.
```
Comment (SimonRichardson) [inline internal/logger/loggo.go:1]: This

**Comment by hmlanigan** [comment]
@SimonRichardson, may or may not be related to your change.
```
$ juju debug-log -m controller --include-labels charmhub
ERROR include label key value "charmhub" not valid
```

**Comment by SimonRichardson** [comment]
> @SimonRichardson, may or may not be related to your change.
> 
> ```
> $ juju debug-log -m controller --include-labels charmhub
> ERROR include label key value "charmhub" not valid
> ```
> 
> Not unique to the charmhub label, metrics is not valid as well. It's only `--include-labels` which has the problem, `juju model-config logging-config="<root>=info;#charmhub=TRACE" -m controller` worked as expected and trace output is in the debug log.

This has changed on main to become `juju debug-log -m controller --include-labels logger-tags=charmhub` - we can discuss in person if we like this or not.

**Review by manadart — APPROVED** [review]
Logging config changes also work.

**Pull request merged**